cogboard-app/src/main/kotlin/com/cognifide/cogboard/logStorage/LogStorage.kt

```diff
@@ -0,0 +1,293 @@
+        if (size > 0 && size > maxSize) {
+            val deleteFactor = ((size - maxSize).toDouble() / size)
+            val logCount = collection.countDocuments()
+            val toDelete = (logCount.toDouble() * deleteFactor).toLong()
+            if (toDelete > 0) {
+                LOGGER.debug("Deleting $toDelete logs as the size $size exceeds maximum size of $maxSize")
+                deleteFirstLogs(toDelete)
+            }
+        }
+    }
+
+    /** Filters in place the logs not matching the rules. */
+    private fun filter(logs: MutableList<Log>) {
+        val regexes = enabledRegexes
+        if (regexes.isEmpty()) { return }
+        logs.retainAll { log ->
+            log.variableData.any { variable ->
+                regexes.any { it.containsMatchIn(variable.header) }
+            }
+        }
+    }
+
+    /** Deletes logs not matching to the rules from the database. */
+    fun filterExistingLogs() {
+        val collection = logsCollection ?: return
+
+        val fieldName = Log.VARIABLE_DATA + "." + LogVariableData.HEADER
+        val regexes = enabledRegexes.map { regex(fieldName, it.pattern) }
+
+        if (regexes.isEmpty()) { return }
+
+        collection.deleteMany(nor(regexes))
+    }
+
+    /** Downloads new logs and inserts the to the database. Returns the number of inserted logs. */
+    private fun downloadInsertLogs(seq: Long, skipFirstLines: Long? = null): Long {
+        var sequence = seq
+        val logs = connection
+                .getLogs(skipFirstLines)
+                .mapNotNull { parserStrategy.parseLine(it) }
+                .toMutableList()
+
+        // Filter the logs by quarantine rules
+        filter(logs)
+        logs.forEach {
+            it.seq = sequence
+            sequence += 1
+        }
+
+        logsCollection?.insertMany(logs.map { it.toDocument() })
+
+        return logs.size.toLong()
+    }
+
+    /** Checks how many logs to download, downloads them and saves them to the database. */
+    private fun downloadLogs() {
+        // Get the current settings
+        val storageConfig = collectionConfiguration
+        var lastLine = storageConfig?.lastLine ?: 0
+        var seq = storageConfig?.seq ?: 0
+
+        // Get the number of lines in the file
+        val fileLineCount = connection.getNumberOfLines() ?: 0
+
+        if (fileLineCount > 0 && fileLineCount > lastLine) {
+            // Download new logs and append them
+            val inserted = downloadInsertLogs(seq, lastLine)
+            lastLine += inserted
+            seq += inserted
+        } else if (fileLineCount in 1 until lastLine) {
+            // Remove all logs and download from the beginning
+            deleteAllLogs()
+            seq = 0
+            val inserted = downloadInsertLogs(seq)
+            lastLine = inserted
+            seq += inserted
+        }
+
+        // Save the new configuration
+        saveConfiguration(LogCollectionConfiguration(config.id, lastLine, seq))
```

Comment:
Isn't `LogCollectionConfiguration` more like a state? Maybe a better name would be `LogCollectionState`

Reply:
Changed.